### ISSUE TITLE
If used with ember-cli-deploy-gzip the uploaded files won't work.

### ISSUE THREAD

**Issue by urbany** [dschmidt/ember-cli-deploy-sentry#26]
Hi, I'm using this addon alongside `ember-cli-deploy-gzip` and since gzip runs first, the files deploy-sentry uploads to sentry are gzipped and won't work properly.

I tried disabling gzip for map files but since the js files are still being gzipped (as I want them to be on my server) this is what I see on my sentry stacktrace:

![screen shot 2016-11-03 at 21 58 24](https://cloud.githubusercontent.com/assets/50884/19988852/9161c85e-a21a-11e6-9a2a-30d41f4b266f.png)

Any idea on how to solve this? Is it possible to check if the file is gzipped before uploading and if so?

1. Add a header to the file upload post which will make sentry properly handle the gzipped file?
2. Or unzip the file before uploading?

Thank you very much in advance

**Comment by urbany** [comment]
I see this issue has been brought up before in #8 and #9 but keeping the zipped files isn't a possibility for me. Any ideas on how to fix this for good?

**Comment by dschmidt** [comment]
I dont have time to look into this right now :-(

**Comment by dschmidt** [comment]
Are you using master or the release version btw?

**Comment by urbany** [comment]
Release. I can submit a PR, just need some guidance on what is the best solution.

**Comment by dschmidt** [comment]
It would be a start to try master, I'm afraid I can't give any guidance atm. I haven't done any Ember or Node.js work lately - you're probably all on your own here.

**Comment by urbany** [comment]
Hehe, ok mate no problem I'll give it a try then.

**Comment by duizendnegen** [comment]
We should also push this convo to `ember-cli-deploy-gzip`, there's cases where we want to keep both the processed files (possibly in `distFiles: []`) and unprocessed files (possibly in `rawFiles: []`).

@urbany there are (from my point of view) 3 options, sorted in order of preference:
1. see if Sentry accepts gzipped assets if we're smarter with headers that we sent to them;
2. see if we can get the raw/unprocessed files from `ember-cli-deploy-gzip`;
3. unzip the files separately again, writing a `ember-cli-deploy-ungzip`. I vaguely recall other plugins have travelled this route, it would not be my preference at all.

**Comment by urbany** [comment]
@duizendnegen thank you for replying, I agree with all your points and priorities.

**Comment by urbany** [comment]
So I just spent a few hours trying to solve this and looking at @duizendnegen priorities:
1. could not make it work, tried using the following headers during upload and it doesn't fix the problem: `Transfer-Encoding: gzip`, `Content-Encoding: gzip`
2. ember-cli-deploy-gzip has an option to keeo the original files, but then the gzipped files have a different name so it's not what we need, and doing something there to keep the originals seems overkill.
3. This is the only option I can see working now, I can take the idea from #9 and adapt to current master if you guys are interested.

Right now I decided to stop GZIPping my files and let cloudfront do it for me.

**Comment by duizendnegen** [comment]
Could you open a ticket on the `Sentry` side of things and see if they're seeing similar issues? According to https://github.com/getsentry/sentry/issues/934 this should just work out of the box (but obviously it doesn't)
Thanks for spending some cycles on investigating!

**Comment by urbany** [comment]
@duizendnegen I opened a ticket yesterday https://github.com/getsentry/sentry/issues/4566 and I'm currently in contact with their support via email. Current status is: it should work but it's not working, so they are going to fix it. I'll keep you posted when I have an update.

**Comment by duizendnegen** [comment]
Any update on this, did support get back to you?

**Comment by urbany** [comment]
@duizendnegen nothing yet, still waiting.

**Comment by kmiyashiro** [comment]
Funny since Sentry explicitly says they do not support gzipped artifacts: https://docs.sentry.io/clients/javascript/sourcemaps/#verify-artifacts-are-not-gzipped

If getsentry/sentry#4566 actually gets fixed and you can send gzipped artifacts with a header, this plugin will have to support setting the header for this to work. As far as I can tell, this is not supported yet: https://github.com/dschmidt/ember-cli-deploy-sentry/blob/master/index.js#L171-L174

**Comment by urbany** [comment]
@kmiyashiro when they fix it I'll add a PR with a configurable option for the user to indicate if the sourcemaps are gzipped, in which case the header will be sent.

PS: thanks for pinging the sentry issue.

**Comment by kmiyashiro** [comment]
@urbany would it be possible to submit the PR before Sentry fixes their side? Since it's supposed to be supported, I don't see why we couldn't add support for it now.

**Comment by kmiyashiro** [comment]
It's happening 🎉  https://github.com/getsentry/sentry/pull/4677

**Comment by duizendnegen** [comment]
Excellent!

fyi we don't need to add an extra configurable option to indicate if the sourcemaps are gzipped, we can simply look at whether `context.gzippedFiles` is present. See https://github.com/ember-cli-deploy/ember-cli-deploy-s3/blob/master/index.js#L32

**Comment by dschmidt** [comment]
Sure, I will accept a PR as soon as it has landed on sentry master. Feel free to send it before the sentry PR is merged.

Great work everyone :+1:

**Comment by kmiyashiro** [comment]
Looks like the sentry PR has been merged.

**Comment by alexjpong** [comment]
So it looks like we're still waiting for sentry to merge the gzip code? In the mean time, would it be best to rely on something like Cloudfront to gzip instead of `ember-cli-deploy-gzip`?

**Comment by duizendnegen** [comment]
That's right @alexjpong. There are alternative paths for when Sentry decides it doesn't want to support gzipping, but it's clumsier and much more work (convincing `-gzip` to keep the non-encrypted files, rename files after uploading to Sentry so that also S3 and azure-blob also find things like they expect...) - I'd rather wait a bit.

**Comment by jpaas** [comment]
This issue hasn't been updated in 6 months. Is there any chance we will be able to use ember-cli-deploy-gzip?

**Comment by urbany** [comment]
I asked about this yesterday: https://github.com/getsentry/sentry/issues/4566
No response yet.

**Comment by kmiyashiro** [comment]
I went the route of unzipping and renaming the unzipped files `*.gz` before uploading them to sentry with an in-repo addon, I couldn't wait for the fix.

**Comment by xcambar** [comment]
@kmiyashiro care to share? This would help at least the 6 contributors of this thread and probably many more!

**Comment by kmiyashiro** [comment]
Sure, here is the in-repo addon, be sure to put this right before the upload to sentry and after you upload your gzipped files where they need to go. In this case, I'm running after `s3` which uploads the gzipped files.

```js
/* eslint-env node */
const fs = require('fs');
const glob = require('glob');

module.exports = {
  name: 'ember-cli-deploy-unzip-sentry',

  createDeployPlugin: function(options) {
    return {
      name: options.name,

      runBefore: 'sentry',
      runAfter: ['gzip', 's3'],

      upload: function() {
        let jsFiles = glob.sync('./dist/**/*.js');

        jsFiles.forEach((file) => {
          let fileData = fs.readFileSync(file, 'utf-8');
          fs.writeFileSync(`${file}.gz`, fileData);
          console.log(`✔ copied ${file} to ${file}.gz`);
        });
      },
    };
  }
};
```
